### PR TITLE
coccinelle: rework how we run the Coccinelle transformations

### PR DESCRIPTION
**Pull request by mrc0mmand**
Turns out that the original way we did things was quite broken, as it
skipped a _lot_ of code. This was because we just threw everything into
one pile and tried to spatch it, but this made Coccinelle sad, like when
man page examples redefined some of our macros, causing typedef
conflicts.

For example, with a minimal reproducer that defines a cleanup macro in
two source files, Coccinelle has no issues when spatch-ing each one
separately:
```
$ spatch --verbose-parsing --sp-file zz-drop-braces.cocci main.c
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: main.c
SPECIAL NAMES: adding _cleanup_ as a attribute with arguments
SPECIAL NAMES: adding _cleanup_free_ as a attribute
```
```
$ spatch --verbose-parsing --sp-file zz-drop-braces.cocci
logcontrol-example.c
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: logcontrol-example.c
SPECIAL NAMES: adding _cleanup_ as a attribute with arguments
```
But when you try to spatch both of them at once, Coccinelle starts
complaining and skipping the "bad" code:
```
$ spatch --verbose-parsing --sp-file zz-drop-braces.cocci main.c logcontrol-example.c
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: main.c logcontrol-example.c
SPECIAL NAMES: adding _cleanup_ as a attribute with arguments
SPECIAL NAMES: adding _cleanup_free_ as a attribute
remapping: _cleanup_ to an ident in macro name
ERROR-RECOV: found sync end of #define, line 44
parsing pass2: try again
ERROR-RECOV: found sync end of #define, line 44
parse error
 = File "logcontrol-example.c", line 44, column 21, charpos = 1719
  around = '__attribute__',
  whole content = #define _cleanup_(f) __attribute__((cleanup(f)))
badcount: 2
bad: #include <systemd/sd-journal.h>
bad:
BAD:!!!!! #define _cleanup_(f) __attribute__((cleanup(f)))
```
This was, unfortunately, hidden as it is visible only with
--verbose-parsing (or --parse-error-msg).

Another issue was how we handled includes. The original way of throwing
them into the pile of source files doesn't really work, leading up to
similar issues as above. The better way is to let Coccinelle properly
resolve all includes by telling it where to find our own include files
(basically the same thing we already do during compilation).

After fixing all this, Coccinelle now has a chance to process much more
of our code (there are still some issues in more complex macros, but
that requires further investigation). However, there's a huge downside
from all of this - doing a _proper_ code analysis is surprisingly time
and resource heavy; meaning that processing just one Coccinelle rule now
takes 15 - 30 minutes.

To make this slightly less painful, Coccinelle supports caching the
generated ASTs, which actually helps a lot - it gets the runtime of one
rule from 15 - 30 minutes down to ~1 minute. It, of course, has its own
downside - the cache is _really_ big (ATTOW the cache takes ~15 GiB).

However, even with the aggressive AST caching you're still looking at
~1 hour for one full Coccinelle run, which is a bit annoying, but I
guess that's the price of doing things _properly_ (but I'll definitely
look into ways of further optimizing this).

---

This was all prompted by @yuwata's https://github.com/systemd/systemd/pull/30617, where I attempted to write an additional rule for `in_addr_hash_func()`, but couldn't, for the love of God, figure out why it kept ignoring valid code.

### CODE REVIEW

File: src/shared/killall.c [hunk 1]
```diff
@@ -257,7 +257,7 @@ static int killall(int sig, Set *pids, bool send_sighup) {
 
                 r = pidref_kill(&pidref, sig);
                 if (r < 0) {
-                        if (errno != -ESRCH)
+                        if (errno != ESRCH)
                                 log_warning_errno(errno, "Could not kill " PID_FMT ", ignoring: %m", pidref.pid);
```
Comment (yuwata) [inline src/shared/killall.c]: No, `pidref_kill()` is our function, hence please use negative errno:
```c
if (r < 0) {
        if (r != -ESRCH)
                log_warning_errno(r, "...");
}
```

File: src/shared/tpm2-util.c [hunk 2]
```diff
@@ -5891,10 +5891,7 @@ int tpm2_unseal_data(
                                        "Failed to unseal data: %s", sym_Tss2_RC_Decode(rc));
 
         _cleanup_(iovec_done) struct iovec d = {};
-        d = (struct iovec) {
-                .iov_base = memdup(unsealed->buffer, unsealed->size),
-                .iov_len = unsealed->size,
-        };
+        d = IOVEC_MAKE(memdup(unsealed->buffer, unsealed->size), unsealed->size);
```
Comment (yuwata) [inline src/shared/tpm2-util.c:5894]: Not you fault, but missing oom check.

File: src/modules-load/modules-load.c [hunk 3]
```diff
@@ -172,10 +172,8 @@ static int run(int argc, char *argv[]) {
                 log_warning_errno(r, "Failed to parse kernel command line, ignoring: %m");
 
         ctx = kmod_new(NULL, NULL);
-        if (!ctx) {
-                log_error("Failed to allocate memory for kmod.");
-                return -ENOMEM;
-        }
+        if (!ctx)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOMEM), "Failed to allocate memory for kmod.");
```
Comment (yuwata) [inline src/modules-load/modules-load.c]: Not a matter, but maybe log_oom() is enough in this case.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks super nice. Thank you! But several comments.

File: src/libsystemd/sd-bus/bus-message.c [hunk 4]
```diff
@@ -627,7 +627,7 @@ static int message_new_reply(
                         return r;
         }
 
-        t->dont_send = !!(call->header->flags & BUS_MESSAGE_NO_REPLY_EXPECTED);
+        t->dont_send = call->header->flags & BUS_MESSAGE_NO_REPLY_EXPECTED;
```
Comment (YHNdnzj) [inline src/libsystemd/sd-bus/bus-message.c]: Maybe just switch to `FLAGS_SET`?

File: src/systemctl/systemctl-edit.c [hunk 5]
```diff
@@ -247,13 +247,12 @@ static int find_paths_to_edit(
...
-                                log_info("Run 'systemctl edit%s --force --full %s' to create a new unit.",
-                                         arg_runtime_scope == RUNTIME_SCOPE_GLOBAL ? " --global" :
-                                         arg_runtime_scope == RUNTIME_SCOPE_USER ? " --user" : "",
-                                         *name);
-                                return -ENOENT;
-                        }
+                        if (!arg_force)
+                                return log_info_errno(SYNTHETIC_ERRNO(ENOENT),
+                                                      "Run 'systemctl edit%s --force --full %s' to create a new unit.",
+                                                      arg_runtime_scope == RUNTIME_SCOPE_GLOBAL ? " --global" :
```
Comment (YHNdnzj) [inline src/systemctl/systemctl-edit.c:253]: While at it, could you please switch this to `runtime_scope_cmdline_option_to_string` too?

File: src/shared/killall.c (hunk 1 above)
Comment (mrc0mmand) [reply src/shared/killall.c]: Ouch, ouch.

File: src/shared/tpm2-util.c (hunk 2 above)
Comment (mrc0mmand) [reply src/shared/tpm2-util.c:5894]: There is an oom check just after the following `explicit_bzero_safe()`.

File: src/systemctl/systemctl-edit.c (hunk 5 above)
Comment (mrc0mmand) [reply src/systemctl/systemctl-edit.c:253]: Wouldn't this then print unnecessary `--system` when using the default (system) scope? I.e.

```
$ build/systemctl edit sdfasdfa
No files found for sdfasdfa.service.
Run 'systemctl edit --system --force --full sdfasdfa.service' to create a new unit.
``` 

I mean, I don't mind it, it's just slightly inconsistent with other places where we try to hide this, like

https://github.com/systemd/systemd/blob/96f69497e316f8ffb8fc8ca0bab9fb5b3b854aa0/src/systemctl/systemctl-edit.c#L93

File: src/shared/tpm2-util.c (hunk 2 above)
Comment (yuwata) [reply src/shared/tpm2-util.c:5894]: Indeed. Sorry for the noise.

File: src/systemctl/systemctl-edit.c (hunk 5 above)
Comment (yuwata) [reply src/systemctl/systemctl-edit.c:253]: Hmm, I like the previous version then...

Comment (mrc0mmand) [reply src/systemctl/systemctl-edit.c:253]: I pushed the previous version again, i.e. the version that only adds `SYNTHETHIC_ERRNO()`. We can follow-up this in another PR once we agree on the format.

Comment (yuwata) [reply src/systemctl/systemctl-edit.c:253]: Yeah, let's do that later.

**Review by yuwata — APPROVED** [review]

**Pull request merged**